### ISSUE TITLE
Multiple-Input models

### ISSUE THREAD

**Issue by cypherex** [keras-rl/keras-rl#254]
Hi,

I already understand the keras functional model, my question is more specific to Keras RL and the structure of data in the Environment objects for models with multiple inputs.

how can I set up a DQN model that has multiple inputs to the neural network (convolutional and a dense input), in Keras-RL. And how would I pass in the "Observation" via the environment when it is more than one input?

I have no issue setting up a model with 1 input.

**Comment by mirraaj** [comment]
I request you to let me know what do you define as input here?

On thing you could do is concatenate all the inputs if it is 1-Dimensional. Alternatively, you can define input in the following manner. 
```
def model_fn(inp, name='inputs'):
 	inps = Input(tensor=inp, name=name)

 	# Define your model here.

 	# Note : Parameter sharing is not working
 	# Hence define two different parallel models
 	# for critic and actor networks.

 	x_actor = Dense(32, activation='relu')(inps)
 	x_actor = Dense(16)(x_actor)
 	x_actor = ReLU(max_value=80.)(x_actor)

 	x_critic = Dense(32, activation='relu')(inps)
 	x_critic = Dense(16, activation='relu')(x_critic)

 	# Actor and Critic output for the model
 	actor_output = Dense(nb_actions, activation='softmax')(x_actor)
 	critic_output = Dense(nb_actions, activation='linear')(x_critic)

 	# Input list to the model
 	inputs = [inps]

 	# Output list to the model
 	outputs = [critic_output, actor_output]

 	model = Model(inputs=inputs, outputs=outputs)
 	return model, inputs, outputs
```
I used this format to define my neural network for ACER algorithm.

**Comment by Sycor4x** [comment]
@mirraaj This seems to work for some models but not for others. For example, the keras-rl DDPG code explicitly assumes that the critic's state inputs are the same as the actors. https://github.com/keras-rl/keras-rl/blob/master/rl/agents/ddpg.py#L126

**Comment by cypherex** [comment]
So that is what I define as an input. My concern is how the states are fed into this network through the keras-rl library, especially with window sizes more than one.

You define the environment to return state, reward and terminal during a step.

So how do I shape my state such that I can have two inputs which are both important to representing the state.

**Comment by mirraaj** [comment]
I think to do so, you should look into https://github.com/keras-rl/keras-rl/blob/master/rl/core.py . Here you may have to define your state as per your definition.

However, I believe that if you are following the gym format, you do not need to add any extra line of code. The gym format itself returns a defined state variable which could be used to train the network.

Please let me know if I failed to give an appropriate answer.

**Comment by bklebel** [comment]
@cypherex if you have not found a solution yet, you may want to look at #195, a pull-request which would enable keras-rl to digest dictionary like observation spaces, where every entry of the dictionary can be a separate state, fed into a separate part of your model. 

The pull-request is not yet fully tested, and progress has slowed down a bit, but I am on it.

**Comment by srivatsankrishnan** [comment]
@mirraaj @bklebel Do you guys know if MultiInputProcessor works with DDPG? If yes, is there an example that you can point me to?
I am running into an issue when using MultiInputProcessor work with DDPG.[Here is the link that describes it in little more depth](https://github.com/keras-rl/keras-rl/issues/287l)

**Comment by bklebel** [comment]
@srivatsankrishnan 
I think if you used named inputs, as described in my pull-request #195 (sadly not yet implemented for ddpg, otherwise I would have run into your problem already),  I think it would be fairly easy to do. I will answer more in depth at your issue #287, possibly later today.
